### ISSUE TITLE
Support running linters on `textDocument/didSave`

### ISSUE THREAD

**Issue by jose-elias-alvarez** [jose-elias-alvarez/null-ls.nvim#166]
#69 describes the issues with `golangci-lint`, a linter that doesn't support `stdin` and also needs to run on a full project for context (meaning temp files won't work unless we create them within the project, which could cause other issues). Due to the timing of `textDocument/didChange` notifications, linting files on disk doesn't provide a good experience (the diagnostics "get stuck" and don't update as expected), which is why the PR is on hold. 

One solution would be to allow linters to only run on `textDocument/didSave` notifications. At the moment, it isn't possible to run both types of linters at the same time, since diagnostics from `didChange` linters continually clobber the ones from `didSave` linters, but neovim/neovim#15585 introduces the idea of diagnostic namespaces, which will solve this issue perfectly. 

The question is then whether supporting this is a good idea. The experience between the two types of linters is inconsistent, and I wonder whether it would cause confusion (or cause users to think that null-ls itself is broken). However, it seems like this isn't a problem for the VS Code ecosystem – for example, as far as I can tell the [golangci-lint integration for VS Code](https://golangci-lint.run/usage/integrations/) only runs on save. 

Since there seems to be demand to run these tools, I think it's a good idea to support them once the Neovim PR is merged, as long as we improve the documentation to avoid unnecessary bug reports / complaints. 

Edit: This could also be relevant to #155, though I think using temp files is always a better option if it's possible to do so.
